### ISSUE TITLE
[PROPOSAL] Generation-visible dynamic tools with measurable prefix reuse

### ISSUE THREAD

**Issue by tpae** [osaurus-ai/vmlx-swift#213]
## Summary

Osaurus tested a KV-preserving capability handoff against `OsaurusAI/Nanbeige4.2-3B-JANG_6M` at vmlx-swift revision `5052be1ad6fdecdbc0da111abf8db5744894d17a`.

- **Parser-only candidate:** keep `UserInput.tools` frozen and expand only `LMInput.toolSchemas` after a capability load. Result: **0/3**. The model called the already prompt-visible `file_write` instead of the newly loaded `eval_dynamic_load_probe`.
- **Expanded-prompt baseline:** include the loaded tool in `UserInput.tools` on the next iteration. Result: **3/3** with `capabilities → eval_dynamic_load_probe → file_write`.
- Both memory-only runs reported `KV +0hit/+0miss`, so partial-prefix reuse was not observable in the current telemetry.

This appears consistent with the current implementation rather than a parser defect: `UserInput.tools` is passed to `applyChatTemplate` and influences generation, while `LMInput.toolSchemas` is consumed later by `ToolCallProcessor` for recognition and validation. `withToolSchemas(...)` does not change prompt tokens or logits.

## Reproduction

The eval performs this sequence in one agent run:

1. Call `capabilities` with `ids: ["tool/eval_dynamic_load_probe"]`.
2. Call `eval_dynamic_load_probe` with `{}`.
3. Call `file_write` to write `loaded.txt` containing exactly `ok`.

```bash
OSAURUS_EVALS_KV_REGIME=memory-only make evals-report \
  EVALS_SUITE=Packages/OsaurusEvals/Suites/AgentLoop \
  FILTER=capabilities-load-midrun \
  MODEL=OsaurusAI/Nanbeige4.2-3B-JANG_6M \
  OSAURUS_EVALS_SKIP_PREP=1 \
  EVALS_FLOOR_FLAG= \
  EVALS_OUT=build/evals/capability-kv-variant.json
```

The candidate attached the expanded schema only after prompt preparation:

```swift
let prepared = try await processor.prepare(input: userInput) // frozen UserInput.tools
let input = prepared.withToolSchemas(expandedTools)           // parser-only expansion
```

Observed candidate sequence in every trial:

```text
capabilities → file_write
```

The model's final prose sometimes claimed that `eval_dynamic_load_probe` ran, but the actual parsed/executed trace showed it did not.

The baseline re-rendered the prompt with the expanded tools and produced the required sequence in all trials:

```text
capabilities → eval_dynamic_load_probe → file_write
```

### 1. Document the two tool contracts explicitly

Please clarify that:

- `UserInput.tools` is generation-visible because it participates in chat-template rendering.
- `LMInput.toolSchemas` is parser-visible and does not teach the model a new tool or constrain sampled tool names.

### 2. Prove or fix partial reuse when prompt tools expand

vmlx-swift already has longest-common-prefix cache logic, but the Nanbeige run exposed no hit/miss delta. It would help to report per request:

- `matchedPrefixTokens`
- `promptTokens`
- `divergenceIndex`
- `restoreTier`
- `rePrefilledTokens`
- a typed reason when reuse is unavailable

A focused topology test could render prompt A with `{capabilities,file_write}`, render prompt B with an added `eval_dynamic_load_probe`, and assert:

- a non-zero partial match;
- never a false full hit;
- correct re-prefill after the divergence point;
- correct advancement of Nanbeige's 44 loop-layer KV slots.

Changing an earlier tool block causally invalidates all later KV states, so the realistic target is exact-prefix restore plus suffix re-prefill—not reuse of the entire old prompt.

### 3. Evaluate an optional generation-visible dynamic-tool API

For compatible tool-call formats, one possible experiment is an opt-in allowed-tool-name trie sourced from `LMInput.toolSchemas`, implemented through the existing stateful `LogitProcessor` hook. It should:

- activate only inside a recognized tool-name span;
- support multi-token names;
- be explicitly gated by `ToolCallFormat`/model family;
- leave argument parsing and schema validation in `ToolCallProcessor`;
- remain experimental, because name constraints alone do not provide tool semantics.

The capability result already contains the loaded tool's description/schema, so this experiment would test whether generation can select a newly described tool without re-rendering the original tool block. It should not be enabled globally or used to coerce unsupported model families.

## Acceptance criteria

- Expanded-prompt behavior remains **3/3** with the exact `capabilities → eval_dynamic_load_probe → file_write` sequence.
- Post-load generation reports a truthful cache result: non-zero exact-prefix match plus suffix re-prefill, or an explicit reason reuse is unavailable.
- If an optional name constraint is implemented, the parser-only experiment becomes **3/3** without substituting a previously prompt-visible tool.
- No raw tool/reasoning markers leak; arguments validate against the expanded schemas; TTFT and token/s are recorded.
- Nanbeige loop-layer offsets remain consistent.
- Existing tool parser, cache topology, canonical-boundary, and Nanbeige runtime suites remain green.

## Environment

- OS: Darwin 25.5.0, Apple Silicon
- Model: `OsaurusAI/Nanbeige4.2-3B-JANG_6M`
- vmlx-swift revision: `5052be1ad6fdecdbc0da111abf8db5744894d17a`
- KV regime: memory-only

## Interim downstream behavior

Osaurus is retaining the expanded prompt schema from osaurus-ai/osaurus#2300 and accepting one post-load re-prefill iteration. The parser-only optimization is considered blocked until generation correctness and cache reuse are both proven.
